feat(zone.js): add a zone config to allow user disable wrapping uncaught promise rejection

JiaLiPassion opened the pull request:
Close #27840.

By default, `zone.js` wrap uncaught promise error and wrap it to a new Error object with some
additional information includes the value of the error and the stack trace.

Consider the following example:

```
Zone.current
  .fork({
    name: 'promise-error',
    onHandleError: (delegate: ZoneDelegate, current: Zone, target: Zone, error: any): boolean => {
      console.log('caught an error', error);
      delegate.handleError(target, error);
      return false;
    }
}).run(() => {
  const originalError = new Error('testError');
  Promise.reject(originalError);
});
```

The `promise-error` zone catches a wrapped `Error` object whose `reject` property equals
to the `originalError`, and the message will be `Uncaught (in promise): testError...`.
You can disable this wrapping behavior by defining a global configuraiton
` __zone_symbol__DISABLE_WRAPPING_UNCAUGHT_PROMISE_REJECTION = true;` before importing `zone.js`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #28740

This PR was original submitted here https://github.com/angular/angular/pull/31443, and it was reverted as it is a breaking change, and in this new PR, I changed the logic and keep the original behavior and use a configuration to let user change the behavior.

mhevery commented on integration/_payload-limits.json:
```diff
@@ -3,8 +3,8 @@
-        "main-es2015": 141569,
-        "polyfills-es2015": 36657
+        "main-es2015": 141943,
+        "polyfills-es2015": 37081
```
Why such large changes in size, given that code changes are relatively minor.

JiaLiPassion replied:
I changed the logic a bit, now the size change will under the limit, so we don't need to update this file for now. And the reason the size is bigger because last time the `Passive event` PR's changes also included.

mhevery commented on packages/zone.js/lib/common/promise.ts:
```diff
@@ -176,18 +178,30 @@ Zone.__load_patch('ZoneAwarePromise', (global: any, Zone: ZoneType, api: _ZonePr
-            error.task = Zone.currentTask !;
-            _uncaughtPromiseErrors.push(error);
+          if (queue.length == 0 && state == REJECTED) {
+            (promise as any)[symbolState] = REJECTED_NO_CATCH;
```
The above two lines are same as the two lines above that. Is that intentional?

JiaLiPassion replied:
Sorry, those lines are not needed, deleted.

mhevery commented on packages/zone.js/lib/common/promise.ts:
```diff
@@ -176,18 +178,30 @@ Zone.__load_patch('ZoneAwarePromise', (global: any, Zone: ZoneType, api: _ZonePr
+                    'Uncaught (in promise): ' + readableObjectToString(value) +
+                    (value && value.stack ? '\n' + value.stack : ''));
+              } catch (err) {
+                uncaughtPromiseError = err;
```
What is the benefit of throwing and than catching the error? If there is a reason, can you document it please so others are not confused like me.

JiaLiPassion replied:
I believe the reason of the logic here is if the `rejected value` is not an `Error` object, we will build an `Error` object to add more readable message and `stack trace`.

mhevery commented on packages/zone.js/lib/zone.configurations.api.ts:
```diff
@@ -530,6 +530,17 @@ interface ZoneGlobalConfigurations {
+  /**
+   * Disable wrapping uncaught promise rejection.
+   *
+   * By default, `zone.js` wrap the uncaught promise rejection to a new `Error` object
```
```suggestion
   * By default, `zone.js` wraps the uncaught promise rejection in a new `Error` object
```

JiaLiPassion replied:
Got it, updated. Thanks.

mhevery commented on packages/zone.js/lib/zone.configurations.api.ts:
```diff
@@ -530,6 +530,17 @@ interface ZoneGlobalConfigurations {
+   * Disable wrapping uncaught promise rejection.
+   *
+   * By default, `zone.js` wrap the uncaught promise rejection to a new `Error` object
+   * with additional information, the value of the rejection and the stack trace.
```
```suggestion
   * which contains additional information such as a value of the rejection and a stack trace.
```

JiaLiPassion replied:
Got it, updated, thanks.

mhevery commented on packages/zone.js/lib/zone.configurations.api.ts:
```diff
@@ -530,6 +530,17 @@ interface ZoneGlobalConfigurations {
+   * with additional information, the value of the rejection and the stack trace.
+   *
+   * If you set `__zone_symbol__DISABLE_WRAPPING_UNCAUGHT_PROMISE_REJECTION = true;` before
+   * importing `zone.js`, `zone.js` does not wrap the uncaught promise rejection.
```
```suggestion
   * importing `zone.js`, `zone.js` will not wrap the uncaught promise rejection.
```

JiaLiPassion replied:
Got it, updated, thanks.

mhevery commented on packages/zone.js/test/common/promise-disable-wrap-uncaught-promise-rejection.spec.ts:
```diff
@@ -0,0 +1,80 @@
+          task = Zone.currentTask;
+          error = new Error('rejectedErrorShouldBeHandled');
+          try {
+            // throw so that the stack trace is captured
```
we don't seem to assert anything about a stack. So this `try-catch` seems unneeded?

JiaLiPassion replied:
The `try/catch` is for `IE` browser, because the `stack` property will only be there when `throw` in IE.

mhevery commented on packages/zone.js/test/common/promise-disable-wrap-uncaught-promise-rejection.spec.ts:
```diff
@@ -0,0 +1,80 @@
+}
+
+describe('disable wrap uncaught promise rejection', () => {
+  it('should notify Zone.onHandleError if no one catches promise', (done) => {
```
```suggestion
  it('should notify Zone.onHandleError if promise is uncaught', (done) => {
```

Is `uncaught` the right word? you mean no one processes rejection??? Or is that the same thing?

JiaLiPassion replied:
I believe `uncaught` and `no one processes rejection` is the same thing here.

mhevery commented on packages/zone.js/test/common/promise-disable-wrap-uncaught-promise-rejection.spec.ts:
```diff
@@ -0,0 +1,80 @@
+    });
+  });
+
+  it('should print original information when throw a not error object with a message property',
```
```suggestion
  it('should print original information when a non-Error object is used for rejection',
```

JiaLiPassion replied:
Got it, updated, thanks.

JiaLiPassion committed:
feat(zone.js): add a zone config to allow user disable wrapping uncaught promise rejection

Close #27840.

By default, `zone.js` wrap uncaught promise error and wrap it to a new Error object with some
additional information includes the value of the error and the stack trace.

Consider the following example:

```
Zone.current
  .fork({
    name: 'promise-error',
    onHandleError: (delegate: ZoneDelegate, current: Zone, target: Zone, error: any): boolean => {
      console.log('caught an error', error);
      delegate.handleError(target, error);
      return false;
    }
}).run(() => {
  const originalError = new Error('testError');
  Promise.reject(originalError);
});
```

The `promise-error` zone catches a wrapped `Error` object whose `rejection` property equals
to the original error, and the message will be `Uncaught (in promise): testError....`,
You can disable this wrapping behavior by defining a global configuraiton
`__zone_symbol__DISABLE_WRAPPING_UNCAUGHT_PROMISE_REJECTION = true;` before importing `zone.js`.

mhevery commented:
[presubmit](http://test/OCL:299849439:BASE:299849484:1583768489601:792a0f63)

AndrewKushnir commented:
FYI, changes in this PR were tested against g3 as a part of https://github.com/angular/angular/pull/36043, using Global TAP run, [which was successful](https://test.corp.google.com/ui#id=OCL:300603599:BASE:301077666:1584325797358:d8800cfc&include_skipped=true). Thank you.
